Python/libraries/recognizers-date-time/recognizers_date_time/date_time/base_date.py

```diff
@@ -0,0 +1,768 @@
+        match = regex.match(self.config.week_day_and_day_of_month_regex, trimmed_source)
+        if match:
+            day_str = match.group('DayOfMonth')
+            er = ExtractResult.get_from_text(day_str)
+            day = int(self.config.number_parser.parse(er).value)
+            month = reference.month
+            year = reference.year
+
+            # the validity of the phrase is guaranteed in the Date Extractor
+            result.timex = FormatUtil.luis_date(year, month, day)
+            date = datetime(year, month, day)
+            result.future_value = date
+            result.past_value = date
+            result.success = True
+
+            return result
+
+        return result
+
+    def parse_weekday_of_month(self, source: str, reference: datetime) -> DateTimeParseResult:
+        trimmed_source = source.strip()
+        result = DateTimeResolutionResult()
+        match = regex.match(self.config.week_day_of_month_regex, trimmed_source)
+        if not match:
+            return result
+        cardinal_str = RegExpUtility.get_group(match, 'cardinal')
+        weekday_str = RegExpUtility.get_group(match, 'weekday')
+        month_str = RegExpUtility.get_group(match, 'month')
+        no_year = False
+        cardinal = 5 if self.config.is_cardinal_last(cardinal_str) else self.config.cardinal_map.get(cardinal_str)
+        weekday = self.config.day_of_week.get(weekday_str)
+        month = reference.month
+        year = reference.year
+
+        if not month_str:
+            swift = self.config.get_swift_month(trimmed_source)
+            temp = reference.replace(month=reference.month + swift)
+            month = temp.month
+            year = temp.year
+        else:
+            month = self.config.month_of_year.get(month_str)
+            no_year = True
+
+        value = self._compute_date(cardinal, weekday, month, year)
+
+        if value.month != month:
+            cardinal -= 1
+            value = value.replace(day=value.day - 7)
+
+        future_date = value
+        past_date = value
+        if no_year and future_date < reference:
+            future_date = self._compute_date(cardinal, weekday, month, year + 1)
+            if future_date.month != month:
+                future_date = future_date.replace(day=future_date.day - 7)
+
+        if no_year and past_date >= reference:
+            past_date = self._compute_date(cardinal, weekday, month, year - 1)
+            if past_date.month != month:
+                past_date = past_date.replace(day=past_date.date - 7)
+
+        result.timex = '-'.join(['XXXX', FormatUtil.to_str(month, 2), 'WXX', str(weekday), '#' + str(cardinal)])
+        result.future_value = future_date
+        result.past_value = past_date
+        result.success = True
+        return result
+
+    def _compute_date(self, cardinal: int, weekday: DayOfWeek, month: int, year: int):
+        first_day = datetime(year, month, 1)
```

Comment:
Why prefix it with '_'? I'm not familiar with this convention.

Reply:
Python does not have the concept of private variables. However, there is a convention to treat functions with an underscore as a non-public functions. [(more information)](https://docs.python.org/3.5/tutorial/classes.html#private-variables)